migrations/helpers/dart_samples_update_helper.py

```diff
@@ -0,0 +1,387 @@
+    end_datetime = datetime.strptime(s_end_datetime, MONGO_DATETIME_FORMAT)
+
+    if start_datetime > end_datetime:
+        print("Aborting run: End datetime must be greater than Start datetime")
+        return
+
+    print(f"Starting DART update process with Start datetime {start_datetime} and End datetime " f"{end_datetime}")
+
+    try:
+        mongo_docs_for_sql = []
+        number_docs_found = 0
+
+        # open connection to mongo
+        with create_mongo_client(config) as client:
+            mongo_db = get_mongo_db(config, client)
+
+            samples_collection = get_mongo_collection(mongo_db, COLLECTION_SAMPLES)
+
+            print(f"Selecting positive samples between {s_start_datetime} and {s_end_datetime}")
+
+            # 1. get samples from mongo between these time ranges that are positive
+            first_match = {
+                "$match": {
+                    # 1. First filter by the start and end dates
+                    FIELD_CREATED_AT: {"$gte": start_datetime, "$lte": end_datetime},
+                    FIELD_RESULT: {"$regex": "^positive", "$options": "i"},
+                    FIELD_ROOT_SAMPLE_ID: {"$not": {"$regex": "^CBIQA_"}},
+                }
+            }
+
+            pipeline = [first_match]
+
+            results = samples_collection.aggregate(pipeline)
+
+            num_docs = 0
+
+            root_sample_ids = []
+            plate_barcodes = set()
+
+            for doc in results:
+                root_sample_ids.append(doc[FIELD_ROOT_SAMPLE_ID])
+                plate_barcodes.add(doc[FIELD_PLATE_BARCODE])
+
+                num_docs += 1
+
+            print(f"{num_docs} to process")
+            print(f"{len(plate_barcodes)} unique plate barcodes")
+
+            # 2. of these, find which have been cherry-picked and remove them from the list
+            cp_samples_df = get_cherrypicked_samples(config, root_sample_ids, plate_barcodes)
+
+            # create a second match for the pipeline
+            second_match = {
+                "$match": {
+                    FIELD_ROOT_SAMPLE_ID: {"$nin": cp_samples_df[FIELD_ROOT_SAMPLE_ID].to_list()},  # type: ignore
+                }
+            }
+
+            pipeline.append(second_match)  # type: ignore
+
+            # get the samples between those dates minus the cherry-picked ones
+            results = samples_collection.aggregate(pipeline)
+            samples = list(results)
+
+            print(
+                f"{len(samples)} documents found in the mongo database between these timestamps and not cherry-picked"
+            )
+
+            # 3. add the relevant filtered positive fields if not present
+            filtered_positive_identifier = FilteredPositiveIdentifier()
+            version = filtered_positive_identifier.current_version()
+            update_timestamp = datetime.now()
+
+            print("Updating filtered positives...")
+
+            update_filtered_positive_fields(
```

Comment:
This will update the filtered positive fields on **all samples**. This should be fine as long as any pending filtered positive rule change migration is run first. If not, one situation in which it might not be okay is for a sample currently being cherrypicked, but not actually picked: you could update this sample to be not eligible for cherrypicking in mongo and MLWH, but this wouldn't go to DART as the sample's plate would be 'started'.

The safest way to do this would be to add filtered positive info if none existed on the sample (e.g. one of the three expected fields doesn't exist), as you know that sample won't be in the process of being cherrypicked. Although I can't think of a non-rare set of circumstances in which this will manifest as a problem, and the one I can think of isn't that much of a problem (picking a sample that has just been updated so it shouldn't be picked).

Will let you decide whether mitigating a small, possibly insignificant issue is worth adding more complexity here. Maybe just a comment?

Reply:
Wow, yes good spot of an edge-case. I think I'll add more comments for now and revisit if time permits.